File: php/reader.php
<?php

require_once 'types.php';

class Reader {
    protected $tokens = array();
    protected $position = 0;
    public function __construct($tokens) {
        $this->tokens = $tokens;
        $this->position = 0;
    }
    public function next() {
        if ($this->position >= count($this->tokens)) { return null; }
        return $this->tokens[$this->position++];
    }
    public function peek() {
        if ($this->position >= count($this->tokens)) { return null; }
        return $this->tokens[$this->position];
    }
}

class BlankException extends Exception {
}

function _real_token($s) {
    return $s !== '' && $s[0] !== ';';
}

function tokenize($str) {
    $pat = "/[\s,]*(~@|[\[\]{}()'`~^@]|\"(?:\\\\.|[^\\\\\"])*\"|;.*|[^\s\[\]{}('\"`,;)]*)/";
    preg_match_all($pat, $str, $matches);
    return array_values(array_filter($matches[1], '_real_token'));
}

function read_atom($reader) {
    $token = $reader->next();
    if (preg_match("/^-?[0-9]+$/", $token)) {
        return intval($token, 10);
    } elseif ($token[0] === "\"") {
        $str = substr($token, 1, -1);
        $str = preg_replace('/\\\\"/', '"', $str);
        return $str;
    } elseif ($token[0] === ":") {
        return _keyword(substr($token,1));
    } elseif ($token === "nil") {
        return NULL;
    } elseif ($token === "true") {
        return true;
    } elseif ($token === "false") {
        return false;
    } else {
        return _symbol($token);
    }
}

function read_list($reader, $constr='_list', $start='(', $end=')') {
    $ast = $constr();
    $token = $reader->next();
    if ($token !== $start) {
        throw new Exception("expected '" . $start . "'");
    }
    while (($token = $reader->peek()) !== $end) {
        if ($token === "" || $token === null) {
            throw new Exception("expected '" . $end . "', got EOF");
        }
        $ast[] = read_form($reader);
    }
    $reader->next();
    return $ast;
}

function read_hash_map($reader) {
    $lst = read_list($reader, '_list', '{', '}');
    return call_user_func_array('_hash_map', $lst->getArrayCopy());
}

function read_form($reader) {
    $token = $reader->peek();
    switch ($token) {
    case '\'': $reader->next();
               return _list(_symbol('quote'),
                               read_form($reader));
    case '`':  $reader->next();
               return _list(_symbol('quasiquote'),
                               read_form($reader));
    case '~':  $reader->next();
               return _list(_symbol('unquote'),
                               read_form($reader));
    case '~@': $reader->next();
               return _list(_symbol('splice-unquote'),
                               read_form($reader));
    case '^':  $reader->next();
               $meta = read_form($reader);
               return _list(_symbol('with-meta'),
                               read_form($reader),
                               $meta);

    case '@':  $reader->next();
               return _list(_symbol('deref'),
                               read_form($reader));

    case ')': throw new Exception("unexpected ')'");
    case '(': return read_list($reader);
    case ']': throw new Exception("unexpected ']'");
    case '[': return read_list($reader, '_vector', '[', ']');
    case '}': throw new Exception("unexpected '}'");
    case '{': return read_hash_map($reader);

    default:  return read_atom($reader);
    }
}

function read_str($str) {
    $tokens = tokenize($str);
    if (count($tokens) === 0) { throw new BlankException(); }
    return read_form(new Reader($tokens));
}

?>

File: matlab/reader.m
% this is just being used as a namespace
classdef reader
    methods (Static = true)
        function tokens = tokenize(str)
            re = '[\s,]*(~@|[\[\]{}()''`~^@]|"(?:\\.|[^\\"])*"|;[^\n]*|[^\s\[\]{}(''"`,;)]*)';
            % extract the capture group (to ignore spaces and commas)
            tokens = cellfun(@(x) x(1), regexp(str, re, 'tokens'));
            comments = cellfun(@(x) length(x) > 0 && x(1) == ';', tokens);
            tokens = tokens(~comments);
        end
      
        function atm = read_atom(rdr)
            token = rdr.next();
            %fprintf('in read_atom: %s\n', token);
            if not(isempty(regexp(token, '^-?[0-9]+$', 'match')))
                atm = str2double(token);
            elseif strcmp(token(1), '"')
                atm = token(2:length(token)-1);
                atm = strrep(atm, '\"', '"');
                atm = strrep(atm, '\n', char(10));
            elseif strcmp(token(1), ':')
                atm = types.keyword(token);
            elseif strcmp(token, 'nil')
                atm = types.nil;
            elseif strcmp(token, 'true')
                atm = true;
            elseif strcmp(token, 'false')
                atm = false;
            else
                atm = types.Symbol(token);
            end
        end

        function seq = read_seq(rdr, start, last)
            %fprintf('in read_seq\n');
            seq = {};
            token = rdr.next();
            if not(strcmp(token, start))
                error(sprintf('expected ''%s''', start));
            end
            token = rdr.peek();
            while true
                if eq(token, false)
                    error(sprintf('expected ''%s''', last));
                end
                if strcmp(token, last), break, end
                seq{end+1} = reader.read_form(rdr);
                token = rdr.peek();
            end
            rdr.next();
        end

        function lst = read_list(rdr)
            seq = reader.read_seq(rdr, '(', ')');
            lst = types.List(seq{:});
        end

        function vec = read_vector(rdr)
            seq = reader.read_seq(rdr, '[', ']');
            vec = types.Vector(seq{:});
        end

        function map = read_hash_map(rdr)
            seq = reader.read_seq(rdr, '{', '}');
            map = types.HashMap(seq{:});
        end

        function ast = read_form(rdr)
            %fprintf('in read_form\n');
            token = rdr.peek();
            switch token
            case ''''
                rdr.next();
                ast = types.List(types.Symbol('quote'), ...
                                 reader.read_form(rdr));
            case '`'
                rdr.next();
                ast = types.List(types.Symbol('quasiquote'), ...
                                 reader.read_form(rdr));
            case '~'
                rdr.next();
                ast = types.List(types.Symbol('unquote'), ...
                                 reader.read_form(rdr));
            case '~@'
                rdr.next();
                ast = types.List(types.Symbol('splice-unquote'), ...
                                 reader.read_form(rdr));
            case '^'
                rdr.next();
                meta = reader.read_form(rdr);
                ast = types.List(types.Symbol('with-meta'), ...
                                 reader.read_form(rdr), meta);
            case '@'
                rdr.next();
                ast = types.List(types.Symbol('deref'), ...
                                 reader.read_form(rdr));

            case ')'
                error('unexpected '')''');
            case '('
                ast = reader.read_list(rdr);
            case ']'
                error('unexpected '']''');
            case '['
                ast = reader.read_vector(rdr);
            case '}'
                error('unexpected ''}''');
            case '{'
                ast = reader.read_hash_map(rdr);
            otherwise
                ast = reader.read_atom(rdr);
            end
        end

        function ast = read_str(str)
            %fprintf('in read_str\n');
            tokens = reader.tokenize(str);
            rdr = types.Reader(tokens);
            ast = reader.read_form(rdr);
        end
    end
end
